tests/unit/test_charm.py

```diff
@@ -275,3 +275,63 @@ def test_unused_unit_sets_waiting_status_on_consumer_joined(self):
             self.harness.model.unit.status,
             WaitingStatus("inactive unit"),
         )
+
+    def test_alert_rules(self):
```

Comment:
Consider having a dedicated test file for this, e.g. `test_relation_data.py`.
Even more so when `begin_with_initial_hooks()` is called in `setUp` because `set_leader` and `add_relation` [should be called beforehand](https://github.com/canonical/operator/blob/ff30d2f4bf0d8e2f24cd11d99e3de67d8e81e3b0/ops/testing.py#L213).

Reply:
The reason this is in test_charm is it is testing basic functionality. I don't really so a reason to rewrite all the boiler plate for something that makes sense to be in here.

Could you explain the issue with `set_leader` and and `add_relation` going after begin? Both of these things can happen after being deployed in production so why is it wrong to do it in test code?

Reply:
There is not one. This is not prescriptive, it's just an example in the docstring for how to hydrate data.

There is absolutely no problem in doing it afterwards to fire events as they come.

Reply:
> set_leader and add_relation should be called beforehand

My bad, seems like that is not the case.

Reply:
I actually think it's a reasonable thought from reading the docstring and a good indication that it could be updated to at least reflect the fact that they are hydrating it, which could be useful in "racing" event firing.